pkg/oauthserver/server/grant/grant.go

```diff
@@ -294,13 +289,13 @@ func (l *Grant) failed(reason string, w http.ResponseWriter, req *http.Request)
 	l.render.Render(form, w, req)
 }
 func (l *Grant) redirect(reason string, w http.ResponseWriter, req *http.Request) {
-	then := req.FormValue("then")
+	then := req.FormValue(thenParam)
 
-	// TODO: validate then
-	if len(then) == 0 {
+	if !redirect.IsServerRelativeURL(then) {
```

Comment:
Why change to IsRelativeServerURL here ?

Reply:
Prevent open redirect on `<master>/oauth/authorize/approve?then=http://evil.com`